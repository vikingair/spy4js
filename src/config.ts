type SpyNonSetupConfig = {
    useOwnEquals: boolean;
    enforceOrder: boolean;
    useGenericReactMocks: boolean;
    afterEachCb?: () => void;
};

export type SpyConfig = SpyNonSetupConfig & {
    afterEach?: (cb: () => void) => void;
    beforeEach?: (cb: () => void) => void;
    expect?: { addSnapshotSerializer: (serializer: any) => void; getState: () => { currentTestName?: string } };
    runner: 'jest' | 'vitest' | 'other';
};

/**
 * Initial default settings for every
 * spy instance. Can be modified only
 * implicitly by "Spy.configure".
 */
const defaults: SpyConfig = {
    useOwnEquals: false,
    enforceOrder: true,
    useGenericReactMocks: false,
    afterEach: (global as any).afterEach,
    beforeEach: (global as any).beforeEach,
    expect: (global as any).expect,
    runner: process.env.JEST_WORKER_ID !== undefined ? 'jest' : 'vitest',
};

export const Config = { ...defaults };

export const configure = (config: Partial<SpyNonSetupConfig>) => configureAll(config);

export const configureAll = (config: Partial<SpyConfig> = {}): void => {
    Object.entries(config).forEach(([name, value]) => {
        if (value !== undefined) {
            Config[name as keyof SpyConfig] = value as never;
        }
    });
};
