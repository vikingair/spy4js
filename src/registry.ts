import { Symbols } from './symbols';

/**
 * This file is part of spy4js which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 */
const restoreAttributeForEntry = (value: any): void => {
    const { obj, method, methodName } = value;
    if (obj) {
        obj[methodName]?.[Symbols.onRestore]?.();
        obj[methodName] = method;
    }
};

type SpyRegister = { [index: number]: { obj: any; method: (...args: any[]) => any; methodName: string } };

export class SpyRegistry {
    register: SpyRegister;
    persReg: SpyRegister;
    registerCount: number;

    /**
     * The SpyRegistry is a class to handle the
     * correct restoration of spied objects.
     *
     * You may push an objects information about
     * a spied method to store it and be able to
     * restore it at any time.
     * Consider the SpyRegistry as information storage
     * for spied objects.
     *
     * @constructor
     */
    constructor() {
        this.register = {};
        this.persReg = {};
        this.registerCount = 0;
    }

    /**
     * If called, the SypRegistry will be resetting to its initial state.
     * Exception: the unique register count will not be touched.
     * Meaning that all stored object information will be restored
     * to their individual previous state.
     */
    restoreAll(): void {
        Object.values(this.register).forEach((entry) => {
            restoreAttributeForEntry(entry);
        });
        this.register = {};
    }

    /**
     * If called, the SpyRegistry will restore the object,
     * which was registered at given index and is getting lose
     * of the stored information.
     *
     * If the registry entry for the given index does not exist,
     * nothing will happen.
     *
     * @param {number} index -> the unique identifier of stored information.
     */
    restore(index: number): void {
        const entry = this.register[index];
        if (entry) {
            restoreAttributeForEntry(entry);
            delete this.register[index];
        }
    }

    /**
     * If called, the SpyRegistry will store the given information.
     * The unique identifier index will be returned.
     *
     * @param obj -> The related object, which will be spied.
     * @param methodName -> The name of the mocked method.
     * @return {number} -> The unique store index.
     */
    push(obj: Record<string, unknown>, methodName: keyof typeof obj): number {
        this.registerCount += 1;
        this.register[this.registerCount] = {
            obj,
            method: obj[methodName] as (...args: any[]) => any,
            methodName,
        };
        return this.registerCount;
    }

    /**
     * If called, the stored method for the corresponding index
     * will be returned. If the registry entry does not exist,
     * undefined will be returned.
     *
     * @param index -> the unique identifier of stored information.
     * @return {any} -> Any stored information can be returned.
     *                   BUT: Usually this method returns a function or
     *                        undefined.
     */
    getOriginalMethod(index: number): ((...args: any[]) => any) | void {
        const entry = this.register[index];
        if (entry) {
            return entry.method;
        }
    }

    /**
     * If called, the stored method will be moved from the standard
     * registry into the persistent registry or vice versa.
     * This can make restore and restoreAll having no effect anymore.
     *
     * @param index -> the unique identifier of stored information.
     * @param intoPersReg -> boolean to determine the moving direction.
     */
    persist(index: number, intoPersReg: boolean): void {
        const fromReg = intoPersReg ? this.register : this.persReg;
        const toReg = intoPersReg ? this.persReg : this.register;
        const entry = fromReg[index];
        if (entry) {
            toReg[index] = entry;
            delete fromReg[index];
        }
    }
}
