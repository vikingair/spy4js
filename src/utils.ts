/**
 * This file is part of spy4js which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 */

import { IGNORE, serialize } from './serializer';

type Comparator = (arg: any) => boolean | void;
type ComparatorOrMapper = (arg: any) => boolean | void | string;
const SPY_COMPARE_FAILED = 'Spy.COMPARE failed';
const SPY_MAPPER_FAILED = 'Spy.MAPPER failed';
/**
 * Uniquely identifiable container for spy relevant comparators.
 */
class SpyComparator {
    _func: ComparatorOrMapper;
    constructor(comparator: ComparatorOrMapper) {
        this._func = comparator;
    }
    compare(arg: any): string[] | undefined {
        const result = this._func(arg);
        if (typeof result === 'string') return [`${SPY_MAPPER_FAILED} [${result}]`];
        if (result === false) return [SPY_COMPARE_FAILED];
        return undefined;
    }
}
/**
 * This function may create individual comparators
 * to define argument based comparison for arbitrary
 * nested objects.
 *
 * Returning "any" because this return value should be usable anywhere.
 */
const COMPARE = (comparator: Comparator): any => new SpyComparator(comparator);

/**
 * Returning "any" because this return value should be usable anywhere.
 */
const MAPPER = (from: any | any[], to: any): any =>
    new SpyComparator((mapper: (...args: any[]) => any) => {
        const result = mapper(...(Array.isArray(from) ? from : [from]));
        const diff = differenceOf(result, to);
        return diff ? `${serialize(result)} did not match ${serialize(to)}: ${diff}` : undefined;
    });

const __different = (type: string) => ['different ' + type];

/**
 * This function is the internal representation of
 * "differenceOf". It does recursively call itself.
 * Read more below.
 *
 * @param {any} a <- any param.
 * @param {any} b <- any param to compare with the first param.
 * @param {boolean} useOwnEquals <-  enables/disables the usage
 *                                   of own "equals" implementations.
 * @param {Array<any>} alreadyComparedArray
 *      <- All flatly compared objects
 *         will be temporarily stored
 *         in this array to resolve
 *         circular structures.
 *
 * @return {string|void} <- information about the difference
 *                           of the provided arguments.
 */
const __diff = (a: any, b: any, useOwnEquals: boolean, alreadyComparedArray: Array<any> = []): string[] | undefined => {
    if (a === IGNORE || b === IGNORE) return undefined;
    if (a instanceof SpyComparator) return a.compare(b);
    if (b instanceof SpyComparator) return b.compare(a);
    if (a === b) return undefined;
    if (a === undefined || b === undefined) return ['one was undefined'];
    if (a === null || b === null) return ['one was null'];
    const aClass = Object.prototype.toString.call(a);
    const bClass = Object.prototype.toString.call(b);
    if (aClass !== bClass) return __different(`object types: ${aClass} <-> ${bClass}`);
    switch (aClass) {
        case '[object RegExp]':
            if (String(a) === String(b)) {
                return undefined;
            }
            return __different('regexp');
        case '[object String]':
            return __different('string');
        case '[object Function]':
            return __different('function');
        case '[object AsyncFunction]':
            return __different('async function');
        case '[object Number]':
            if (isNaN(a) && isNaN(b)) {
                return undefined;
            }
            return __different('number');
        case '[object BigInt]':
            return __different('BigInt');
        case '[object Date]':
            if (Number(a) === Number(b)) {
                return undefined;
            }
            return __different('date');
        case '[object Boolean]':
            return __different('bool');
        case '[object Symbol]':
            return __different('symbols');
        case '[object Error]':
            if (String(a) === String(b)) {
                return undefined;
            }
            return __different('error');
        default:
            if (a.constructor !== b.constructor) {
                return __different('constructor');
            }
    }
    if (useOwnEquals && 'equals' in a && typeof a.equals === 'function') {
        if (a.equals(b)) {
            return undefined;
        }
        return [
            'own equals method failed <- ' +
                'Maybe you want to disable the usage ' +
                'of own equals implementation? ' +
                '[ Use: spy.configure({useOwnEquals: false}) ]',
        ];
    }
    if (alreadyComparedArray.indexOf(a) !== -1) {
        return undefined;
    }
    const compared = [...alreadyComparedArray, a];
    const keys = Array.from(new Set(Object.keys(a).concat(Object.keys(b))));
    for (const key of keys) {
        const diff = __diff(a[key], b[key], useOwnEquals, compared);
        if (diff !== undefined) {
            return [key, ...diff];
        }
    }
    return undefined;
};

const __serializeDifferentProp = (obj: any, diff: string[]): string => {
    // asserting diff.length > 1
    let toSerialize = obj;
    diff.slice(0, -1).forEach((key: string) => {
        toSerialize = toSerialize[key];
    });
    return serialize(toSerialize);
};

const __diffToStr = (diff: string[], info: string = ''): string => {
    if (diff.length === 1) return diff[0];
    return `--> ${diff[0]} / ` + diff.slice(1).join(' / ') + info;
};

/**
 * This function does make a comparison of two provided params.
 *
 * If found any difference it will return a string containing
 * information about the first detected difference.
 *
 * If the given params were proven "identical" this function
 * returns undefined.
 *
 * In first place the params are flatly checked. The checks are
 * in the following order:
 *
 * - identical objects
 * - one param is null or undefined
 * - object type
 * - regexp/string/number/date/boolean
 * - constructor
 * - keys length
 * - own "equals" implementation
 *
 * In second place all keys of the params are deeply checked.
 * Using the above flat checks for all attributes.
 *
 * Also all circular structures will be resolved and repeating
 * attributes will be assumed to be equal.
 *
 * @param {any} a <- any param.
 * @param {any} b <- any param to compare with the first param.
 * @param {{useOwnEquals:boolean}} config <- controls the usage of own
 *                                           "equals" implementations.
 *
 * @return {string|void} <- information about the difference
 *                           of the provided arguments.
 */
const differenceOf = (
    a: any,
    b: any,
    config: { useOwnEquals: boolean } = { useOwnEquals: true }
): string | undefined => {
    const diff = __diff(a, b, config.useOwnEquals);
    if (!diff) return undefined;
    const diffStr = __diffToStr(diff);
    if (diff.length < 2) return diffStr;
    const diffProp1 = __serializeDifferentProp(a, diff);
    const lastPart = diff[diff.length - 1];
    if (lastPart.indexOf(SPY_MAPPER_FAILED) === 0) return diffStr;
    const info =
        lastPart === SPY_COMPARE_FAILED
            ? `called with: ${diffProp1}`
            : `${diffProp1} != ${__serializeDifferentProp(b, diff)}`;
    return `${diffStr} [${info}]`;
};

export type MessageOrError = string | Error;
export type OptionalMessageOrError = MessageOrError | undefined;
const toError = (msgOrError: OptionalMessageOrError, spyName: string) =>
    msgOrError instanceof Error ? msgOrError : new Error(msgOrError || `${spyName} was requested to throw`);

export { COMPARE, differenceOf, MAPPER, toError };
