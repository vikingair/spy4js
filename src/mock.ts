/**
 * This file is part of spy4js which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 */
import { Config } from './config';
import { Symbols } from './symbols';

export type Mockable = Record<string, any>;

// returns a spy instance
type SpyOn = (obj: Mockable, method: keyof typeof obj) => any;

const uninitialized = (method: keyof any) => () => {
    throw new Error(`Method '${String(method)}' was not initialized on Mock.`);
};

export const createMock = <T extends Mockable, K extends keyof T>(
    obj: T,
    methods: K[],
    spyOn: SpyOn,
    callsFactory?: MockInfo['callsFactory']
): { [P in K]: any } => {
    const mock = registerMock(obj, spyOn, callsFactory) as { [P in K]: any };
    methods.forEach((method) => {
        mock[method] = uninitialized(method);
    });
    return mock;
};

type MockInfo = {
    mock: Mockable;
    mocked: Mockable;
    callsFactory?: (methodName: string) => (...args: any[]) => any;
    activeMethods: Set<string>;
};

const registerMock = (mocked: Mockable, spyOn: SpyOn, callsFactory?: MockInfo['callsFactory']) => {
    const mock = {};
    const { beforeEach, expect } = Config;

    if (!beforeEach || !expect) throw new Error('You need to call Spy.setup() in order to use mocks.');
    const { currentTestName } = expect.getState();
    if (currentTestName) throw new Error('Mocks can only be created outside of tests');

    const activeMethods = new Set<string>();
    beforeEach(() => {
        initMock({ mocked, mock, callsFactory, activeMethods }, spyOn);
    });

    return mock;
};

export const couldNotInitError = (additional: string) => new Error(`Could not initialize mock because:\n${additional}`);

const initMock = (mockInfo: MockInfo, spyOn: SpyOn): void => {
    const { mocked, mock, callsFactory, activeMethods } = mockInfo;
    Object.keys(mock).forEach((method) => {
        if (activeMethods.has(method)) return;
        try {
            const spy = spyOn(mocked, method as keyof typeof mock);
            mockInfo.activeMethods.add(method);
            if (callsFactory) {
                spy.calls(callsFactory(method));
                spy.displayName = method; // TODO: test if name works, too
            }
            spy[Symbols.onRestore] = () => {
                mockInfo.activeMethods.delete(method);
            };
            mock[method as keyof typeof mock] = spy;
        } catch (e) {
            let msg = (e as Error).message;
            const utilName = Config.runner === 'jest' ? 'jest' : Config.runner === 'vitest' ? 'vi' : undefined;
            if (utilName && msg.includes('has only a getter')) {
                const actual =
                    utilName === 'jest'
                        ? `() => ({ ...jest.requireActual('<module-name>') })`
                        : `async () => ({ ...((await vi.importActual('<module-name>')) as any) })`;
                msg += `
Inserting a module mock might should resolve this problem. Run this code beforehand:

${utilName}.mock('<module-name>');

Or if you don't want to mock everything from this module, you can use this:

${utilName}.mock('<module-name>', ${actual});`;
            }
            throw couldNotInitError(msg);
        }
    });
};
