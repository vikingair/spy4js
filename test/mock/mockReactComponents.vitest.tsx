import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import React from 'react';
import { Spy } from '../../src/spy';
vi.mock('./test.reactComponents', async () => ({ ...((await vi.importActual('./test.reactComponents')) as any) }));

import { render } from '@testing-library/react';
import { _GenericComponent } from '../../src/react';
import { Component1, Component2 } from './test.reactComponents';
import { Component3 } from './test.reactComponents2';

Spy.setup({ expect, beforeEach, afterEach });

describe('mockReactComponents - minimal', async () => {
    Spy.configure({ useGenericReactMocks: false });
    const Mock$TestReactComponents = Spy.mockReactComponents(
        await import('./test.reactComponents'),
        'Component1',
        'Component2'
    );

    it('mocks as plain function', () => {
        expect(Component1({ foo: 'bar' })).toBe(null);
        Mock$TestReactComponents.Component1.wasCalledWith({ foo: 'bar' });
    });

    it('renders component snapshot - nested mocks', () => {
        const { container } = render(
            <Component2>
                <Component1 foo={'bar'} />
            </Component2>
        );
        expect(container).toMatchSnapshot();
    });

    it('renders component snapshot - contained mock', () => {
        const { container } = render(<Component3 />);
        expect(container).toMatchSnapshot();
    });
});

describe('mockReactComponents - generic', async () => {
    beforeEach(() => {
        _GenericComponent.serializeAllProps = true;
    });
    Spy.configure({ useGenericReactMocks: true });
    const Mock$TestReactComponents = Spy.mockReactComponents(
        await import('./test.reactComponents'),
        'Component1',
        'Component2'
    );

    it('mocks as plain function', () => {
        expect((Component1({ foo: 'bar' }) as any).props['data-prop-foo']).toBe("'bar'");
        expect(Mock$TestReactComponents.Component1.getProps().foo).toBe('bar');
        Mock$TestReactComponents.Component1.wasCalledWith({ foo: 'bar' });
    });

    it('renders component snapshot - nested mocks', () => {
        const { container } = render(
            <Component2>
                <Component1 foo={'bar'} oneMore={{ data: Symbol('oneMore'), elem: <div>Test</div> }} />
            </Component2>
        );
        expect(container).toMatchSnapshot();
    });

    it('renders component snapshot - nested mocks - without props', () => {
        _GenericComponent.serializeAllProps = false;
        const { container } = render(
            <Component2>
                <Component1 foo={'bar'} oneMore={{ data: Symbol('oneMore'), elem: <div>Test</div> }} />
                <div>More children</div>
            </Component2>
        );
        expect(container).toMatchSnapshot();
    });

    it('renders component snapshot - contained mock', () => {
        const { container } = render(<Component3 />);
        expect(container).toMatchSnapshot();
    });
});
