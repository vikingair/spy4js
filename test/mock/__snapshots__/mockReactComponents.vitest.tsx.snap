// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mockReactComponents - generic > renders component snapshot - contained mock 1`] = `
<div>
  <div
    id="component-3"
  >
    <samp
      data-component="Component1"
      data-prop-foo="'in-component-3'"
    />
  </div>
</div>
`;

exports[`mockReactComponents - generic > renders component snapshot - nested mocks - without props 1`] = `
<div>
  <samp
    data-component="Component2"
  >
    <samp
      data-component="Component1"
    />
    <div>
      More children
    </div>
  </samp>
</div>
`;

exports[`mockReactComponents - generic > renders component snapshot - nested mocks 1`] = `
<div>
  <samp
    data-component="Component2"
  >
    <samp
      data-component="Component1"
      data-prop-foo="'bar'"
      data-prop-onemore="{data: Symbol(oneMore), elem: <div>Test</div>}"
    />
  </samp>
</div>
`;

exports[`mockReactComponents - minimal > renders component snapshot - contained mock 1`] = `
<div>
  <div
    id="component-3"
  />
</div>
`;

exports[`mockReactComponents - minimal > renders component snapshot - nested mocks 1`] = `<div />`;
